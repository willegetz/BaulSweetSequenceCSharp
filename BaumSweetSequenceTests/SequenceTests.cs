using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Generic;
using ApprovalTests.Reporters;
using ApprovalTests;

namespace BaumSweetSequenceTests
{
    [TestClass]
    [UseReporter(typeof(BeyondCompare4Reporter))]
    public class SequenceTests
    {
        // Convert number to binary :: Binary output is a string
        // Drop preceeding 0's :: Happens on conversion to binary
        //  Minimum number of digits is 1
        // Find all sequences of 0's
        // Does any sequence contain an odd number of 0's?
        //  If Yes: This is a value of 1
        //  If No: This is a value of 0
        // Absence of any 0's in a binary value is considered an absence of any 0 sequence of odd value
        // Iteration of numbers begins at number 0 and loops until reaching the final number

        [TestMethod]
        public void BinaryConversionTest()
        {
            var startingValue = 1;

            var actual = Convert.ToString(startingValue, 2);

            var expected = "1";
            Assert.AreEqual(expected, actual);
        }

        [TestMethod]
        public void Number2HasOdd0Sequence()
        {
            var startingValue = 2;

            var binaryConversion = GetBinaryValue(startingValue);
            var sequencesOf0 = GetBlocksOf0s(binaryConversion);
            var baumSweetEvaluation = GetBaumSweetEvaluation(sequencesOf0);

            var expected = 0;

            Assert.AreEqual(expected, baumSweetEvaluation, "Evaluation contains no block of consecutive 0s of odd length");
        }

        [TestMethod]
        public void Number1DoesNotHaveOdd0Sequence()
        {
            var startingValue = 1;

            var binaryConversion = GetBinaryValue(startingValue);
            var sequencesOf0 = GetBlocksOf0s(binaryConversion);
            int baumSweetEvaluation = GetBaumSweetEvaluation(sequencesOf0);

            var expected = 1;

            Assert.AreEqual(expected, baumSweetEvaluation, "Evaluation contains 1 or more blocks of consecutive 0s of odd length");
        }

        [TestMethod]
        public void BaumSweetSequenceFor2Is1and1and0()
        {
            var baumSweetSequence = new List<int>() { 1 };

            var numberToEvaluate = 2;

            for (int i = 1; i <= numberToEvaluate; i++)
            {
                var binaryConversion = GetBinaryValue(i);
                var sequenceOf0s = GetBlocksOf0s(binaryConversion);
                var sequenceEvaluation = GetBaumSweetEvaluation(sequenceOf0s);

                baumSweetSequence.Add(sequenceEvaluation);
            }

            Approvals.VerifyAll(baumSweetSequence, "b_");
        }
        
        private string GetBinaryValue(int startingValue)
        {
            return Convert.ToString(startingValue, 2);
        }

        private string[] GetBlocksOf0s(string binaryConversion)
        {
            return binaryConversion.Split(new char[] { '1' }, StringSplitOptions.RemoveEmptyEntries);
        }

        private int GetBaumSweetEvaluation(string[] sequencesOf0)
        {
            var hasOdd0Sequence = false;
            var baumSweetEvaluation = 1;

            foreach (var seq in sequencesOf0)
            {
                if (seq.Length % 2 == 1)
                {
                    hasOdd0Sequence = true;
                }
            }

            if (hasOdd0Sequence)
            {
                baumSweetEvaluation = 0;
            }

            return baumSweetEvaluation;
        }
    }
}
