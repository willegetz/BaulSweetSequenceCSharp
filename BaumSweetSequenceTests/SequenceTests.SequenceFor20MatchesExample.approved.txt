b_[0] = 1
b_[1] = 1
b_[2] = 0
b_[3] = 1
b_[4] = 1
b_[5] = 0
b_[6] = 0
b_[7] = 1
b_[8] = 0
b_[9] = 1
b_[10] = 0
b_[11] = 0
b_[12] = 1
b_[13] = 0
b_[14] = 0
b_[15] = 1
b_[16] = 1
b_[17] = 0
b_[18] = 0
b_[19] = 1
b_[20] = 0
